File: src/lib/sammgmtjni/com/sun/netstorage/samqfs/mgmt/stg/Stager.java
/*
 *    SAM-QFS_notice_begin
 *
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at pkg/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at pkg/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 2008 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 *
 *    SAM-QFS_notice_end
 */

// ident	$Id: Stager.java,v 1.13 2008/05/16 18:35:30 am143972 Exp $

package com.sun.netstorage.samqfs.mgmt.stg;

import com.sun.netstorage.samqfs.mgmt.Ctx;
import com.sun.netstorage.samqfs.mgmt.SamFSException;
import com.sun.netstorage.samqfs.mgmt.arc.BufDirective;
import com.sun.netstorage.samqfs.mgmt.arc.DrvDirective;

public class Stager {

    /*
     * options masks for stage files must match the ST_OPT flags
     * in pub/mgmt/stage.h
     */
    public static final int RESET_DEFAULTS = 0x0001;
    public static final int NEVER = 0x0002;
    public static final int ASSOCIATIVE = 0x0004;
    public static final int PARTIAL = 0x0008;
    public static final int RECURSIVE = 0x0010;
    public static final int COPY_1 = 0x1000;
    public static final int COPY_2 = 0x2000;
    public static final int COPY_3 = 0x4000;
    public static final int COPY_4 = 0x8000;

    public static native StagerParams getParams(Ctx c)
        throws SamFSException;
    public static native StagerParams getDefaultParams(Ctx c)
        throws SamFSException;
    public static native void setParams(Ctx c, StagerParams stg)
        throws SamFSException;

    public static native void resetBufDirective(Ctx c, BufDirective bufDir)
        throws SamFSException;
    public static native void resetDrvDirective(Ctx c, DrvDirective drvDir)
        throws SamFSException;

    public static native void idle(Ctx c)
        throws SamFSException;
    public static native void run(Ctx c)
        throws SamFSException;

    /*
     *
     * stage files
     * Used to stage files and directories and to set
     * stage attributes for files and directories.
     *
     * The elements in the lists are matched by index. So the 5th
     * copy entry is used for the fifth filepath entry and the 5th options
     * entry is applied to the staging.
     *
     * In order to signal that a file specified as stage -a be set to
     * stage -n the RESET_DEFAULTS flag must be set also and vice versa.
     *
     * The copies array indicates what copy will be staged from.
     * If the value is 0 the system picks the copy otherwise a valid
     * copy number in the range 1-4 can be specified.
     *
     * Filepaths must contain fully qualified paths.
     *
     * Flags for setting stager options
     * NEVER
     * ASSOCIATIVE
     * RESET_DEFAULTS
     * RECURSIVE
     *
     * Flags for controlling staging
     * PARTIAL
     * RECURSIVE
     *
     * @return null if stage_files successfully completes or an activity id for
     * a SAMASTAGEFILES activity if stage_files is still running.
     *
     */
    public static native String stageFiles(Ctx c, String[] filepaths,
        int options)
        throws SamFSException;


    /**
     * get names of files in staging queue
     *
     * @param mediaType
     * @param vsn
     *
     * @return String[] array of file names
     */
    public static native String[] getFileNamesInQueue(Ctx c, String mediaType,
        String vsn) throws SamFSException;
}

File: src/lib/samut/stager.c
/*
 * stager.c - Stager communication functions.
 */

/*
 *    SAM-QFS_notice_begin
 *
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at pkg/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at pkg/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 *
 *    SAM-QFS_notice_end
 */

#pragma ident "$Revision: 1.20 $"

/* Feature test switches. */

/* ANSI C headers. */
#include <errno.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* POSIX headers. */
#include <sys/types.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/stat.h>

/* Solaris headers. */
#include <sys/mman.h>

/* SAM-FS headers. */
#include "aml/stager.h"
#include "sam/mount.h"
#include "sam/lib.h"
#include "sam/udscom.h"

/* Private data. */
static char ClientName[16] = "libsamut";
static struct UdsClient clnt = { STAGER_SERVER_NAME, ClientName,
	STAGER_SERVER_MAGIC, NULL };

/*
 * Send filesystem mount message to stager.
 */
int
_StagerFsMount(
	const char *SrcFile,
	const int SrcLine,
	char *fsname)
{
	struct StagerFsname arg;
	struct StagerGeneralRsp rsp;
	int status;

	strncpy(ClientName, program_name, sizeof (ClientName)-1);
	strncpy(arg.fsname, fsname, sizeof (arg.fsname)-1);

	status = UdsSendMsg(SrcFile, SrcLine, &clnt, SSR_fs_mount,
	    &arg, sizeof (arg), &rsp, sizeof (rsp));

	if (status == -1) {
		return (-1);
	}
	if (rsp.status == -1) {
		errno = rsp.error;
		return (-1);
	}

	return (0);
}

/*
 * Send filesystem umount message to stager.
 */
int
_StagerFsUmount(
	const char *SrcFile,
	const int SrcLine,
	char *fsname)
{
	struct StagerFsname arg;
	struct StagerGeneralRsp rsp;
	int status;

	strncpy(ClientName, program_name, sizeof (ClientName)-1);
	strncpy(arg.fsname, fsname, sizeof (arg.fsname)-1);

	status = UdsSendMsg(SrcFile, SrcLine, &clnt, SSR_fs_umount,
	    &arg, sizeof (arg), &rsp, sizeof (rsp));

	if (status == -1) {
		return (-1);
	}
	if (rsp.status == -1) {
		errno = rsp.error;
		return (-1);
	}

	return (0);
}

/*
 * Stop removable media activity.  Sent from sam-amld when
 * a 'samd stop' is issued.
 */
int
_StagerStopRm(
	const char *SrcFile,
	const int SrcLine)
{
	struct StagerGeneralRsp rsp;
	int status;

	strncpy(ClientName, program_name, sizeof (ClientName)-1);

	status = UdsSendMsg(SrcFile, SrcLine, &clnt, SSR_stop_rm,
	    NULL, 0, &rsp, sizeof (rsp));

	if (status == -1) {
		return (-1);
	}
	if (rsp.status == -1) {
		errno = rsp.error;
		return (-1);
	}

	return (0);
}

/*
 * Send staging event for log file message to stager.
 */
int
_StagerLogEvent(
	const char *SrcFile,
	const int SrcLine,
	int type,
	int id)
{
	struct StagerLogEvent arg;
	struct StagerGeneralRsp rsp;
	int status;

	strncpy(ClientName, program_name, sizeof (ClientName)-1);
	arg.type = type;
	arg.id = id;

	status = UdsSendMsg(SrcFile, SrcLine, &clnt, SSR_log_event,
	    &arg, sizeof (arg), &rsp, sizeof (rsp));

	if (status == -1) {
		return (-1);
	}
	if (rsp.status == -1) {
		errno = rsp.error;
		return (-1);
	}

	return (0);
}

/*
 * Stager control.
 * Set stager control 'ident' to 'value'.
 * 'ident' is a '.' separated sequence of names of stager data.
 * 'msg' is a buffer to which messages are returned.
 * If 'msg' is NULL, a static buffer is used.
 */
char *
_StagerControl(
	const char *SrcFile,
	const int SrcLine,
	char *ident,
	char *value,
	char *msg,
	int msgsize)
{
	struct StagerControlRsp rsp;
	struct StagerControl arg;
	int status;

	if (ident == NULL || *ident == '\0') {
		strncpy(rsp.msg, "NULL ident", sizeof (rsp.msg)-1);
	} else {
		strncpy(ClientName, program_name, sizeof (ClientName) - 1);
		strncpy(arg.ident, ident, sizeof (arg.ident) - 1);
		if (value != NULL) {
			strncpy(arg.value, value, sizeof (arg.value) - 1);
		} else {
			*arg.value = '\0';
		}
		status = UdsSendMsg(SrcFile, SrcLine, &clnt, SSR_control,
		    &arg, sizeof (arg), &rsp, sizeof (rsp));
		if (status == -1) {
			(void) StrFromErrno(errno, rsp.msg, sizeof (rsp.msg));
		}
	}
	if (msg != NULL) {
		if (msgsize > 0) {
			strncpy(msg, rsp.msg, msgsize - 1);
		} else {
			*msg = '\0';
		}
	} else {
		msg = rsp.msg;
	}
	return (msg);
}
